source/dub/packagesuppliers/registry.d

```diff
@@ -57,19 +57,16 @@ class RegistryPackageSupplier : PackageSupplier {
 			return;
 		auto vers = best["version"].get!string;
 		auto url = m_registryUrl ~ NativePath(PackagesPath~"/"~packageId~"/"~vers~".zip");
-		logDiagnostic("Downloading from '%s'", url);
-		foreach(i; 0..3) {
-			try{
-				download(url, path);
-				return;
-			}
-			catch(HTTPStatusException e) {
-				if (e.status == 404) throw e;
-				else {
-					logDebug("Failed to download package %s from %s (Attempt %s of 3)", packageId, url, i + 1);
-					continue;
-				}
-			}
+		try {
+			retryDownload(url, path);
+			return;
+		}
+		catch(HTTPStatusException e) {
+			if (e.status == 404) throw e;
+			else logDebug("Failed to download package %s from %s", packageId, url); 
```

Comment:
this code can't trigger, if a HTTPStatusException is ever thrown it's only a 404 one (because you catch the rest). Only at the end you throw a normal Exception. Maybe rethrow the last HTTPStatusException on failure after 3 retries instead?

Reply:
HTTPStatusException is now rethrown
The throw Exception on last line of retryDownload I kept in, because of the CurlException